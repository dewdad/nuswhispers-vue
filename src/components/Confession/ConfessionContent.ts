import flow from 'lodash/flow';
import Vue from 'vue';
import Component from 'vue-class-component';
import escapeHtml from '../../utils/escape-html';

const linkTags = (content: string): string => {
  const tags = content.split(/(#\w+)/);
  let formatted = '';

  tags.map((tag) => {
    formatted += /(#\w+)/.test(tag) ?
      `<a href="/tag/${tag.substring(1)}">${tag}</a>` :
      tag;
  });

  return formatted;
};

const linkUrls = (content: string): string => {
  const regex = /(\b(https?|ftp):\/\/[A-Z0-9+&@#\/%?=~_|!:,.;-]*[-A-Z0-9+&@#\/%=~_|])/gim;
  if (content.match(regex)) {
    content = content.replace(regex, '<a rel=\"noopener noreferrer\" href=\"$1\" target=\"_blank\">$1</a>');
  }
  return content;
};

@Component({
  props: {
    content: String,
  },
})
export default class ConfessionContent extends Vue {
  public content: string;

  get formattedContent(): string {
    return flow(escapeHtml, linkTags, linkUrls)(this.content);
  }
}
