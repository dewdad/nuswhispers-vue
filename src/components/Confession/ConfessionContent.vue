<template>
  <p v-html="formattedContent" />
</template>

<script lang="ts" src="./ConfessionContent.ts" />
