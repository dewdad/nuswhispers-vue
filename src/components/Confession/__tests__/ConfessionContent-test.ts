import Vue from 'vue';
import { createRenderer } from 'vue-server-renderer';
import ConfessionContent from '../ConfessionContent.vue';

describe('ConfessionContent', () => {
  const renderer = createRenderer();
  const Ctor = Vue.extend(ConfessionContent);

  it('renders correctly with normal text', () => {
    const vm = new Ctor({
      propsData: {
        content: 'Hello World!',
      },
    });

    renderer.renderToString(vm, (err: Error | undefined, html: string) => {
      expect(html).toMatchSnapshot();
    });
  });

  it('renders correctly with tags', () => {
    const vm = new Ctor({
      propsData: {
        content: 'Hello! #firstworldproblems',
      },
    });

    renderer.renderToString(vm, (err: Error | undefined, html: string) => {
      expect(html).toMatchSnapshot();
    });
  });

  it('renders correctly with HTML tags', () => {
    const vm = new Ctor({
      propsData: {
        content: 'Hello! #foo <script>console.log("YOLO");</script>',
      },
    });

    renderer.renderToString(vm, (err: Error | undefined, html: string) => {
      expect(html).toMatchSnapshot();
    });
  });

  it('renders correctly with URLs', () => {
    const vm = new Ctor({
      propsData: {
        content: 'Hello! https://www.google.com.sg/',
      },
    });

    renderer.renderToString(vm, (err: Error | undefined, html: string) => {
      expect(html).toMatchSnapshot();
    });
  });
});
