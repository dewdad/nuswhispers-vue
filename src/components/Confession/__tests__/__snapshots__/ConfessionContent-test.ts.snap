// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfessionContent renders correctly with HTML tags 1`] = `<p data-server-rendered="true">Hello! <a href="/tag/foo">#foo</a> &lt;script&gt;console.log(&quot;YOLO&quot;);&lt;/script&gt;</p>`;

exports[`ConfessionContent renders correctly with URLs 1`] = `<p data-server-rendered="true">Hello! <a rel="noopener noreferrer" href="https://www.google.com.sg/" target="_blank">https://www.google.com.sg/</a></p>`;

exports[`ConfessionContent renders correctly with normal text 1`] = `<p data-server-rendered="true">Hello World!</p>`;

exports[`ConfessionContent renders correctly with tags 1`] = `<p data-server-rendered="true">Hello! <a href="/tag/firstworldproblems">#firstworldproblems</a></p>`;
